### PR TITLE
Performance: add io_binding support for bert benchmark util

### PR DESCRIPTION
**Pull request by ytaous**
**Description**: Add io_binding support for bert_perf_test.py

**Motivation and Context**
In the recent benchmark for 1p model, we found there's perf gap between python util (bert_bert_test.py) and c++ based util (onnxruntime_perf_test). ~ (10 ms vs. 5.5 ms)

Sample cmd:
ORT_TENSORRT_INT8_ENABLE=0 ORT_TENSORRT_ENGINE_CACHE_ENABLE=1 ORT_TENSORRT_MAX_WORKSPACE_SIZE=4294967296 ORT_TENSORRT_FP16_ENABLE=1 python bert_perf_test.py --model xyz.onnx --batch_size 32 --sequence_length 128 --intra_op_num_threads 24 --use_gpu --provider tensorrt --opt_level=0 --input_ids_name input_ids --input_mask_name attention_mask

./onnxruntime_perf_test -i "trt_engine_cache_enable|true trt_fp16_enable|true" -t 50 -e tensorrt xyz.onnx -s ./res.txt

After experimenting io_binding for the python util, we are able to bring the number down to 
4.02 ms (not output copy to cpu)
7.22 ms (plus output copy to cpu)

Thus we 'd like to add the --use_io_binding support to bring the perf number comparable to c++ based benchmark util.

### CODE REVIEW

File: onnxruntime/python/tools/transformers/bert_perf_test.py [hunk 1]
```diff
@@ -119,6 +121,57 @@ def create_session(model_path, use_gpu, provider, intra_op_num_threads, graph_op
...
+    input_tensors, output_tensors = create_input_output_tensors(inputs, outputs, device)
+    io_binding = create_io_binding(session, input_tensors, output_tensors)
+
+    # warm up
+    for iter in range(1):
+        session.run_with_iobinding(io_binding)
+
+    results = []
+    latency_list = []
+    max_iters = 10
```
Comment (zhangyaobit) [inline onnxruntime/python/tools/transformers/bert_perf_test.py:168]: Should this be set to the length of all_inputs?

Comment (ytaous) [reply onnxruntime/python/tools/transformers/bert_perf_test.py:168]: this is just a repeat tests for the same input from line #152, so 10 times should be good enough.

Comment (zhangyaobit) [reply onnxruntime/python/tools/transformers/bert_perf_test.py:168]: Currently this number of test cases could be decided by a user by using flag --samples (it defaults to 10 indeed), different numbers of samples will affect the statistics computation in run_one_test. Larger number sample will give more accurate statistics, but also more costly to run. 

Although IO binding will just bind to a random test case, I think we should still measure/gather a user-specified number of samples to generate the statistics.

Comment (zhangyaobit) [reply onnxruntime/python/tools/transformers/bert_perf_test.py:168]: Actually, maybe the correct way to do this is to bind for each test case (put io binding in the measurement for loop), but just measure the time session.run_with_iobinding(io_binding)

Comment (ytaous) [reply onnxruntime/python/tools/transformers/bert_perf_test.py:168]: let me wrap it in a loop

File: onnxruntime/python/tools/transformers/bert_perf_test.py [hunk 2]
```diff
@@ -119,6 +121,55 @@ def create_session(model_path, use_gpu, provider, intra_op_num_threads, graph_op
...
+        for i in range(len(output_names)):
+            outputs[output_names[i]] = result[i]
+
+        input_tensors, output_tensors = create_input_output_tensors(inputs, outputs, device)
+        io_binding = create_io_binding(session, input_tensors, output_tensors)
+
+        # warm up
+        session.run_with_iobinding(io_binding)
+
+        for i in range(test_setting.test_times):
```
Comment (zhangyaobit) [inline onnxruntime/python/tools/transformers/bert_perf_test.py:166]: Similar to onnxruntime_inference does, can we move the test_setting.test_times loop outside of this function, in run_one_test?

File: onnxruntime/python/tools/transformers/bert_perf_test.py [hunk 3]
```diff
@@ -119,6 +121,55 @@ def create_session(model_path, use_gpu, provider, intra_op_num_threads, graph_op
...
+                                tensor.data_ptr())
+    return io_binding
+
+def onnxruntime_inference_with_io_binding(session, all_inputs, output_names, test_setting):
+    results = []
+    latency_list = []
+    device = 'cuda' if test_setting.use_gpu else 'cpu'
+    for test_case_id, inputs in enumerate(all_inputs):
+        result = session.run(output_names, inputs)
+        outputs = {}
```
Comment (zhangyaobit) [inline onnxruntime/python/tools/transformers/bert_perf_test.py:156]: results.append(result) ?

Comment (ytaous) [reply onnxruntime/python/tools/transformers/bert_perf_test.py:156]: results is not used, but I can add it to be the same as the other function below.

File: onnxruntime/python/tools/transformers/bert_perf_test.py (hunk 2 above)
Comment (ytaous) [reply onnxruntime/python/tools/transformers/bert_perf_test.py]: sure, will make it comparable, although I think both logic would serve the same purpose

**Review by tianleiwu — APPROVED** [review]

File: onnxruntime/python/tools/transformers/bert_perf_test.py [hunk 4]
```diff
@@ -159,9 +209,14 @@ def run_one_test(model_setting, test_setting, perf_results, all_inputs, intra_op
     print("Running test:", key)
 
     all_latency_list = []
-    for i in range(test_setting.test_times):
-        results, latency_list = onnxruntime_inference(session, all_inputs, output_names)
-        all_latency_list.extend(latency_list)
+    if test_setting.use_io_binding:
```
Comment (zhangyaobit) [inline onnxruntime/python/tools/transformers/bert_perf_test.py:212]: Nit: if else could be inside the for loop to reduce code duplication.

Comment (ytaous) [reply onnxruntime/python/tools/transformers/bert_perf_test.py:212]: good point, perhaps next time whoever touch the code, we'll remind it.

**Pull request merged**